ignite/engine/engine.py

```diff
@@ -1,24 +1,24 @@
-import functools
 import logging
 import math
 import time
 import warnings
-import weakref
 from collections import OrderedDict, defaultdict
 from collections.abc import Mapping
-from typing import Any, Callable, Dict, Iterable, Iterator, List, Optional, Tuple, Union
+from typing import Any, Callable, Iterable, Iterator, List, Optional, Tuple, Union
 
 from torch.utils.data import DataLoader
 
-from ignite.base import Serializable
-from ignite.engine.events import CallableEventWithFilter, EventEnum, Events, EventsList, RemovableEventHandle
+from ignite.base.events import EventEnum, RemovableEventHandle
+from ignite.base.events_driven import EventsDriven
+from ignite.base.mixins import Serializable
+from ignite.engine.events import Events
 from ignite.engine.state import State
 from ignite.engine.utils import _check_signature, _to_hours_mins_secs
 
 __all__ = ["Engine"]
 
 
-class Engine(Serializable):
+class Engine(Serializable, EventsDriven):
```

Comment:
Can't we now inherit from `EventsDrivenWithState` ?

Reply:
Yes but we can't remove the property because of the setter, I will inherit from `EvetsDrivenWithState` but not changes in the code, after forbidding setting state we can remove the property and the setter in `Engine`.